### ISSUE TITLE
Should API handle invalid ids/detail pages?

### ISSUE THREAD

**Issue by hamishwillee** [mdn/express-locallibrary-tutorial#24]
The locallibrary uses ids to refer to particular instances of each "thing". So for example, you can access an author like this: `http://.../catalog/author/5a20cfb5cfc6cb76b0c59207`. The code searches for the id, and either displays an error or renders the page.
```javascript
exports.author_detail = function(req, res, next) {
    async.parallel({
        author: function(callback) {
            Author.findById(req.params.id)
              .exec(callback)
        },
        authors_books: function(callback) {
          Book.find({ 'author': req.params.id },'title summary')
          .exec(callback)
        },
    }, function(err, results) {
        if (err) { return next(err); }
        //Successful, so render
        res.render('author_detail', { title: 'Author Detail', author: results.author, author_books: results.authors_books } );
    });
};
```
The rub is that I just discovered that the `find` method above gives you an error for badly formatted queries, but returning nothing is not an error. The result being that an invalid id will just be displayed as some sort of gibberish.
Now you should never get to that sort of page unless you are manually playing with URLs. So the question for you @chrisdavidmills , is should we fix this. IMO no, but I figured I'd ask -)

If we were to fix this, then there are several options. The first is just to check this case and redirect to the normal error page:
```
    }, function(err, results) {
        if (err) { return next(err); }
        //Successful, so render
        if (results.author==null) {
            var err = new Error('Author Not Found');
            err.status = 404;
            next(err);
        }
        else {
            res.render('author_detail', { title: 'Author Detail', author: results.author, author_books: results.authors_books } );
        }
```
Alternative is to render a custom 404 page. I'd prefer to simply treat this as an error. This is supposed to be simple :-)

**Comment by chrisdavidmills** [comment]
I would like to see some kind of fix for this; I reckon the simpler, first option is OK, and then maybe leave installing a custom error page as a reader exercise?

**Comment by hamishwillee** [comment]
Cheers. Fixed in #29
